core/lib/zksync_core/src/metadata_calculator/recovery.rs

```diff
@@ -0,0 +1,617 @@
+    fn recovery_started(&mut self, chunk_count: usize, recovered_chunk_count: usize) {
+        self.chunk_count = chunk_count;
+        *self.recovered_chunk_count.get_mut() = recovered_chunk_count;
+        RECOVERY_METRICS
+            .recovered_chunk_count
+            .set(recovered_chunk_count);
+    }
+
+    async fn chunk_recovered(&self) {
+        let recovered_chunk_count = self.recovered_chunk_count.fetch_add(1, Ordering::SeqCst) + 1;
+        RECOVERY_METRICS
+            .recovered_chunk_count
+            .set(recovered_chunk_count);
+        let health = Health::from(HealthStatus::Ready).with_details(RecoveryMerkleTreeInfo {
+            mode: "recovery",
+            chunk_count: self.chunk_count,
+            recovered_chunk_count,
+        });
+        self.inner.update(health);
+    }
+}
+
+#[derive(Debug)]
+struct RecoveryOptions<'a> {
+    chunk_count: usize,
+    events: Box<dyn HandleRecoveryEvent + 'a>,
+}
+
+impl GenericAsyncTree {
+    /// Ensures that the tree is ready for the normal operation, recovering it from a Postgres snapshot
+    /// if necessary.
+    pub async fn ensure_ready(
+        self,
+        pool: &ConnectionPool,
+        stop_receiver: &watch::Receiver<bool>,
+        health_updater: &HealthUpdater,
+    ) -> anyhow::Result<Option<AsyncTree>> {
+        let (tree, l1_batch) = match self {
+            Self::Ready(tree) => return Ok(Some(tree)),
+            Self::Recovering(tree) => {
+                let l1_batch = snapshot_l1_batch(pool).await?.context(
+                    "Merkle tree is recovering, but Postgres doesn't contain snapshot L1 batch",
+                )?;
+                let recovered_version = tree.recovered_version();
+                anyhow::ensure!(
+                    u64::from(l1_batch.0) == recovered_version,
+                    "Snapshot L1 batch in Postgres ({l1_batch}) differs from the recovered Merkle tree version \
+                     ({recovered_version})"
+                );
+                tracing::info!("Resuming tree recovery with snapshot L1 batch #{l1_batch}");
+                (tree, l1_batch)
+            }
+            Self::Empty { db, mode } => {
+                if let Some(l1_batch) = snapshot_l1_batch(pool).await? {
+                    tracing::info!(
+                        "Starting Merkle tree recovery with snapshot L1 batch #{l1_batch}"
+                    );
+                    let tree = AsyncTreeRecovery::new(db, l1_batch.0.into(), mode);
+                    (tree, l1_batch)
+                } else {
+                    // Start the tree from scratch. The genesis block will be filled in `TreeUpdater::loop_updating_tree()`.
+                    return Ok(Some(AsyncTree::new(db, mode)));
+                }
+            }
+        };
+
+        // FIXME: make choice of ranges more intelligent. NB: ranges must be the same for the entire recovery!
+        let recovery_options = RecoveryOptions {
+            chunk_count: 256,
```

Comment:
I thought about configuring the desired number of entries in a chunk, but that'd require to get the total number of storage logs in a snapshot. @RomanBrodetski Would a query like this be efficient enough?

```sql
SELECT COUNT(*) FROM storage_logs WHERE miniblock_number = $1;
```

As we've discussed offline, there's no way to precisely reproduce it on the existing DBs since we don't have snapshot miniblocks. A query like this

```sql
select count(*) from storage_logs where miniblock_number between 18000000 and 20000000
```

executes in ~5s on the mainnet.

Reply:
my understanding is that it will have a similar performance to your second query.  You are probably in a better position to judge Whether this is efficitent enough.